File: src/cli/welcome.ts
import chalk from 'chalk'
import { ConfigManager } from '../shared/config.js'
import { BABEL_O_VERSION } from '../shared/version.js'
import { padToTerminalWidth, truncateToTerminalWidth, visibleTerminalWidth } from './terminalWidth.js'
import { getTheme } from './theme.js'

const WELCOME_MAX_WIDTH = 96
const PIXEL_ROWS = [
  '    M    ',
  '   M M   ',
  '    R    ',
  '   R R   ',
  '  R   R  ',
  'O O P V V',
]

const COLORS: Record<string, string> = {
  M: '#ff006e',
  P: '#ff4f9a',
  R: '#c72d68',
  O: '#ff7a18',
  V: '#8b5cf6',
}

function renderLogoRow(row: string): string {
  let result = ''
  for (const char of row) {
    if (char === ' ') {
      result += ' '
    } else {
      const color = COLORS[char] ?? '#ff006e'
      result += chalk.hex(color)('█')
    }
  }
  return result
}

export function renderWelcome(options: {
  modelId?: string
  cwd: string
  sessionId?: string
  url?: string
  title?: string
}): void {
  console.log()
  for (const line of formatWelcomeCardLines(options)) {
    console.log(`  ${line}`)
  }
  console.log()
}

export function formatWelcomeCardLines(options: {
  modelId?: string
  cwd: string
  sessionId?: string
  url?: string
  title?: string
  columns?: number
}): string[] {
  const columns = Math.max(48, options.columns ?? process.stdout.columns ?? 80)
  const maxContentWidth = Math.max(46, Math.min(WELCOME_MAX_WIDTH, columns - 4))
  const mode = options.url ? `Service (${options.url})` : 'Embedded (Local)'
  const username = process.env.USER || process.env.USERNAME || 'User'
  const configManager = ConfigManager.getInstance()
  const defaultModel = options.modelId || configManager.resolveSettings().modelId || 'local/coding-runtime'
  // `session` is a snapshot at the moment the welcome card is
  // rendered. Callers (chat.ts) must invoke renderWelcome AFTER
  // the session id is generated so the card shows the real id
  // rather than the `new session` placeholder. When the
  // optional sessionId is omitted (e.g. for a /config card),
  // fall back to the placeholder so the layout doesn't shift.
  const sessionLabel = options.sessionId
    ? truncateToTerminalWidth(shortenSessionId(options.sessionId), 22)
    : 'new session'
  const logoWidth = Math.max(...PIXEL_ROWS.map(row => visibleTerminalWidth(renderLogoRow(row))))
  const theme = getTheme()
  const metadataWidth = Math.max(18, maxContentWidth - logoWidth - 5)
  const title = options.title ?? `v${BABEL_O_VERSION}`
  // Labeled rows: each metadata line starts with a short grey
  // label so the operator can read the card at a glance.
  // `user` / `model` / `cwd` / `session` / `mode` mirror the
  // /status output the user sees mid-session.
  const labelStyle = chalk.dim
  const metadataLines = [
    ` ${theme.brand('❖ BABEL-O')}  ${chalk.dim(title)}`,
    ` ${labelStyle('user')}    ${chalk.bold.cyan(truncateToTerminalWidth(username, metadataWidth))}`,
    ` ${labelStyle('model')}   ${chalk.yellow(truncateToTerminalWidth(defaultModel, metadataWidth))}`,
    ` ${labelStyle('cwd')}     ${chalk.italic.white(truncateToTerminalWidth(formatCwd(options.cwd), metadataWidth))}`,
    ` ${labelStyle('session')} ${theme.accent(sessionLabel)}`,
    ` ${labelStyle('mode')}    ${theme.accent(truncateToTerminalWidth(mode, metadataWidth))}`,
  ]
  const contentWidths = PIXEL_ROWS.map((row, index) => {
    const logoCol = renderLogoRow(row)
    const metaCol = metadataLines[index] ?? ''
    return visibleTerminalWidth(` ${logoCol}   ${metaCol}`)
  })
  const contentWidth = Math.min(maxContentWidth, Math.max(55, ...contentWidths))
  const lines = []

  for (let i = 0; i < PIXEL_ROWS.length; i++) {
    const logoCol = renderLogoRow(PIXEL_ROWS[i]!)
    const metaCol = metadataLines[i] ?? ''
    const content = ` ${logoCol}   ${metaCol}`
    lines.push(padToTerminalWidth(content, contentWidth))
  }

  return lines
}

export function formatWelcomeHintLine(columns = process.stdout.columns ?? 80): string {
  const width = Math.max(48, Math.min(WELCOME_MAX_WIDTH, columns))
  const left = `${chalk.dim('?')} ${chalk.dim('shortcuts')} ${chalk.dim('·')} ${chalk.dim('/')} ${chalk.dim('commands')} ${chalk.dim('·')} ${chalk.dim('Ctrl+E')} ${chalk.dim('editor')}`
  const right = `${chalk.dim('Ctrl+O')} ${chalk.dim('details')} ${chalk.dim('·')} ${chalk.dim('Ctrl+C')} ${chalk.dim('cancel')}`
  const gap = Math.max(2, width - visibleTerminalWidth(left) - visibleTerminalWidth(right))
  return `${left}${' '.repeat(gap)}${right}`
}

export function formatSessionBanner(action: 'started' | 'resuming', sessionId: string): string {
  const label = action === 'started' ? 'session' : 'resume'
  return `${chalk.dim(label)} ${chalk.dim(sessionId)}`
}

function formatCwd(cwd: string): string {
  const home = process.env.HOME
  if (home && cwd === home) return '~'
  if (home && cwd.startsWith(`${home}/`)) return `~/${cwd.slice(home.length + 1)}`
  return cwd
}

// shortenSessionId renders a session id in the same compact
// form the Go TUI uses (8 chars + "..." + 6 chars tail) so the
// welcome card and the transcript row reference the same
// canonical short id. Falls back to the full id when it's
// already short enough that no truncation is needed.
function shortenSessionId(id: string): string {
  if (id.length <= 18) return id
  return `${id.slice(0, 8)}…${id.slice(-6)}`
}

File: clients/go-tui/internal/tui/welcome.go
package tui

import (
	"os"
	"strings"

	"charm.land/lipgloss/v2"
)

func (m model) renderWelcomeCard(width int) string {
	formattedCwd := m.cfg.Cwd
	home := os.Getenv("HOME")
	if home != "" {
		if formattedCwd == home {
			formattedCwd = "~"
		} else if strings.HasPrefix(formattedCwd, home+"/") {
			formattedCwd = "~/" + formattedCwd[len(home)+1:]
		}
	}

	mode := "Embedded (Local)"
	if m.cfg.BaseURL != "" && !strings.Contains(m.cfg.BaseURL, "127.0.0.1") && !strings.Contains(m.cfg.BaseURL, "localhost") {
		mode = "Service (" + m.cfg.BaseURL + ")"
	}

	defaultModel := m.modelID
	if defaultModel == "" {
		defaultModel = "local/coding-runtime"
	}

	sessionVal := m.sessionID
	if sessionVal == "" {
		sessionVal = m.cfg.SessionID
	}
	if sessionVal == "" {
		sessionVal = "new session"
	}

	pixelRows := []string{
		"    M    ",
		"   M M   ",
		"    R    ",
		"   R R   ",
		"  R   R  ",
		"O O P V V",
	}

	colors := map[rune]string{
		'M': "#ff006e",
		'P': "#ff4f9a",
		'R': "#c72d68",
		'O': "#ff7a18",
		'V': "#8b5cf6",
	}

	renderLogoRow := func(row string) string {
		var sb strings.Builder
		for _, r := range row {
			if r == ' ' {
				sb.WriteByte(' ')
			} else {
				hex, ok := colors[r]
				if !ok {
					hex = "#ff006e"
				}
				style := lipgloss.NewStyle().Foreground(lipgloss.Color(hex))
				sb.WriteString(style.Render("█"))
			}
		}
		return sb.String()
	}

	modelSummary := truncatePlain(defaultModel, 42)
	cwdSummary := truncatePlain(formattedCwd, 42)
	sessionSummary := truncatePlain(shortID(sessionVal), 42)
	modeSummary := truncatePlain(mode, 42)
	titleLine := titleStyle.Render("v"+Version) + mutedStyle.Render("  Welcome back!")
	authValue := "ready"
	authMarker := statusStyle.Render("●")
	if m.authMode != "" && m.authMode != "none" && !m.hasAPIKey {
		authValue = "setup /model"
		authMarker = errorStyle.Render("!")
	}
	metadataLines := []string{
		titleLine,
		welcomeMetaLine("model", modelSummary, statusStyle.Render("●")),
		welcomeMetaLine("auth", authValue, authMarker),
		welcomeMetaLine("session", sessionSummary, mutedStyle.Render("•")),
		welcomeMetaLine("mode", modeSummary, mutedStyle.Render("•")),
		welcomeMetaLine("work", cwdSummary, contextStyle.Render("○")),
	}
	infoGap := 10
	infoOffset := 1 + lipgloss.Width(pixelRows[0]) + infoGap

	var cardLines []string
	for i := 0; i < 6; i++ {
		logoCol := renderLogoRow(pixelRows[i])
		metaCol := ""
		if i < len(metadataLines) {
			metaCol = metadataLines[i]
		}
		combined := " " + logoCol + strings.Repeat(" ", infoGap) + metaCol
		cardLines = append(cardLines, combined)
	}

	maxCardWidth := 0
	for _, line := range cardLines {
		w := lipgloss.Width(line)
		if w > maxCardWidth {
			maxCardWidth = w
		}
	}

	navLine := strings.Join([]string{
		lipgloss.NewStyle().Foreground(lipgloss.Color("#ff7a18")).Render("◆") + " chat",
		lipgloss.NewStyle().Foreground(lipgloss.Color("51")).Render("◆") + " nexus",
		lipgloss.NewStyle().Foreground(lipgloss.Color("99")).Render("◆") + " config",
		lipgloss.NewStyle().Foreground(lipgloss.Color("#ff4f9a")).Render("◆") + " models",
	}, "   ")
	cardLines = append(cardLines, strings.Repeat(" ", infoOffset)+focusedLineStyle.Render(navLine))
	maxCardWidth = max(maxCardWidth, lipgloss.Width(cardLines[len(cardLines)-1]))

	hPad := max(0, (width-maxCardWidth)/2)
	hSpace := strings.Repeat(" ", hPad)

	var outputLines []string
	outputLines = append(outputLines, "", "")
	for _, line := range cardLines {
		outputLines = append(outputLines, hSpace+line)
	}
	outputLines = append(outputLines, "")

	return strings.Join(outputLines, "\n")
}

func welcomeMetaLine(label string, value string, marker string) string {
	labelStyle := mutedStyle.Width(8)
	valueStyle := contextStyle
	if marker == "" {
		marker = mutedStyle.Render("•")
	}
	return marker + " " + labelStyle.Render(label) + " " + valueStyle.Render(value)
}
